_blogposts/2023-05-17-enhancing-the-ergonomics-of-records.mdx

```diff
@@ -0,0 +1,140 @@
+
+```rescript
+type c = {
+  id: string,
+  name: string,
+  age: int,
+  active: bool,
+}
+```
+
+Keeping it as straightforward as possible, spreads are essentially a 'copy-paste' operation for fields from one or more records to another, inlining the fields from the spread records into the new record.
+
+This is going to be a much more ergonomic experience when working with types with many fields, where variations of the same underlying type are needed.
+
+### Use case: extending the built in DOM nodes
+
+This feature can be particularly useful when extending DOM nodes. For instance, in the case of the animation library Framer Motion, one could easily extend the native DOM types with additional properties specific to the library, leading to a more seamless and type-safe integration.
+
+This is how you could bind to a `div` in Framer Motion with the new record type spreads:
+
+```rescript
+type animate = {} // definition omitted for brevity
+
+type divProps = {
+  ...JsxDOM.domProps,
+  initial?: animate,
+  animate?: animate,
+  whileHover?: animate,
+  whileTap?: animate,
+}
+
+module Div = {
+  @module("framer-motion") external make: divProps => Jsx.element = "div"
+}
+```
+
+You can now use `<Div />` as a `<motion.div />` component from Framer Motion. And your type definition is now simple and easy to maintain.
+
+## Type Coercion
+
+Type coercion introduces an extra layer of flexibility when working with records. Now, records of the same shape can be coerced between each other. This means that we can cast record `a` to be record `b` at the type level, if they contain the same exact fields. An example:
+
+```rescript
+type a = {
+  name: string,
+  age: int,
+}
+
+type b = {
+  name: string,
+  age: int,
+}
+
+let nameFromB = (b: b) => b.name
+
+let a: a = {
+  name: "Name",
+  age: 35,
+}
+
+let name = nameFromB((a :> b))
+```
+
+Notice how we _coerce_ the value `a` into the type `b`. This works because they have the same fields.
+
+Additionally, we can also coerce between records where record `a` has a subset of the fields of record `b`. The same example as above, slightly altered:
+
+```rescript
+type a = {
+  id: string,
+  name: string,
+  age: int,
+  active: bool,
+}
+
+type b = {
+  name: string,
+  age: int,
+}
+
+let nameFromB = (b: b) => b.name
+
+let a: a = {
+  id: "1",
+  name: "Name",
+  age: 35,
+  active: true,
+}
+
+let name = nameFromB((a :> b))
+```
+
+Notice how `a` now has more fields than `b`, but we can still coerce `a` to `b` because `b` has a subset of the fields of `a`.
+
+### Coercing is explicit
+
+To maintain the robustness of our type system, coercing is an explicit action. Records are still nominal, preserving the same guarantees as before. You can't accidentally pass record `a` as a `b` without explicitly coercing between them. This is crucial as it prevents accidental dependencies on shapes rather than records, ensuring type safety by forcing explicit coercion between types.
+
+## Conclusion
```

Comment:
I guess records are now in a similar situation as variants. They are vastly superior for error messages and pattern matching ergonomics, but they couple modules and you'd need to bring the name into the module scope so you'd be able to infer types (less flexible, more rigid). Other than that, they are almost on par with their structural counter parts; even the interop part is configurable enough.

Reply:
I'll try and come up with a clever way to incroporate structural objects since they're of course relevant, but I'd like to avoid putting to much emphasis on them. I don't personally view objects and records as "choose either or", because objects are so far behind on usefulness, in my opinion. But I'll find a way to work it in without giving it too much room.

Reply:
okay looks good!